### ISSUE TITLE
`PYI034`: false positive, and quick fix breaks types

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ruff#24462]
## [Playground](https://play.ruff.rs/d79d3fbe-e5c2-472a-abcb-7c79602dcb8a)

```py
class A:
    def __iadd__(self, other: int) -> int:  # `__iadd__` methods in classes like `A` usually return `self` at runtime
        return 1
```

```toml
[lint]
select = [ "PYI034" ]
```

this function usually returns `1` at runtime, not `self`, so the error and the quickfix are not correct

if the intent of the rule is to suggest returning `self` instead of any other type for in-place operators, then the wording should be updated accordingly

**Comment by MichaReiser** [comment]
The in-place operations are special-cased, and the only time that explicitly tests whether the return type isn't `self` instead of testing whether it matches the class's name (`A` in this case). 

Strictly enforcing that all in-place methods return `self` feels more opinionated to me and should probably be its own rule, if that's in fact something we want to enforce. The reason why I say it's more strict is that Python's documentation explicitly mentions that returning `Self` is only recommended but not required:

>  These methods should attempt to do the operation in-place (modifying self) and return the result (which could be, but does not have to be, self). ([source](https://docs.python.org/3/reference/datamodel.html#object.__iadd__))


If the rule's intention is to enforce returning `Self`, then it should also flag the following, which it is not (note the missing return type annotation):

```py
class A:
    def __iadd__(self, other: int):  
        return 1
```

Because of that, I think we should change the condition, even if it means deviating from upstream. But I'd love to hear @AlexWaygood's opinion, because he added support for in-place operations upstream in https://github.com/PyCQA/flake8-pyi/pull/180

**Comment by jonathandung** [comment]
See also [PyCQA/flake8-pyi#539](https://github.com/PyCQA/flake8-pyi/issues/539)

**Comment by AlexWaygood** [comment]
The intent of this rule isn't to ban in-place dunder methods from returning things other than `Self`. It's to catch a common error in stub files where these methods are annotated incorrectly: they nearly always return `self` at runtime, so they should therefore nearly always be annotated with `Self` in stub files.

I think the issue here is that flake8-pyi only supports stub files (and likely only ever will), whereas Ruff's adapted version of the rule also runs on `.py` files. For `.pyi` files, the heuristic makes sense IMO because these in-place dunders really do almost always return `self` at runtime, and in a stub file you haven't really got any other information to go on. I'm not sure this part of `PYI034` really makes sense at all in a `.py` file, however (not the way it's currently written, at least). In a `.py` file, a type checker will quickly tell you off if you write a method like this:

```py
class A:
    def __iadd__(self) -> int:
        return self
```

so in a `.py` file, I think it's probably only useful to emit this diagnostic if the return type references the enclosing class, like with the other parts of PYI034 -- i.e., restrict the rule to complaining about things like this:

```py
class A:
    def __iadd__(self) -> A:
        return self
```

**Comment by KotlinIsland** [comment]
is there any value in reporting:
```py
class A:
    def f(self) -> A:  # use `Self` instead of `A`
        return self
```
